### ISSUE TITLE
Feature Request: Data source to get API Gateway REST API by name

### ISSUE THREAD

**Issue by timborrowdale** [terraform-providers/terraform-provider-aws#4146]
### Use case: 
We would like to be able to independently deploy resource integrations to a single API Gateway. e.g.
1. Main pipeline creates API Gateway
2. Resource 1 pipeline creates lambda function and sets up integration on API Gateway for resource `/1`
3. Resource 2 pipeline creates lambda function and sets up integration on API Gateway for resource `/2`

In order to do this the following terraform resources need knowledge of the existing API Gateway:
* [aws_api_gateway_resource](https://www.terraform.io/docs/providers/aws/r/api_gateway_resource.html)
* [aws_api_gateway_method](https://www.terraform.io/docs/providers/aws/r/api_gateway_method.html)
* [aws_api_gateway_integration](https://www.terraform.io/docs/providers/aws/r/api_gateway_integration.html)

### Suggested data source

```hcl
data "aws_api_gateway_rest_api" "by_name" {
  name = "my-gateway"
}
```

which would return `id` and `root_resource_id` as computed attributes

I can have a go at implementing this if it is considered useful, but I have a concern that currently it is possible to create two API Gateways with the same name. What would be the expected output of a query where there is more than one match?

**Comment by catsby** [comment]
Seems possible, unfortunately there is no "filter" for the [GetRestAPIs](https://docs.aws.amazon.com/apigateway/api-reference/resource/rest-apis/) endpoint. You send the query and get a paginated result, we'd have to loop through and pattern match on the name 😦 

Still, a reasonable request. I don't believe anyone on the Terraform team will be able to get to this, but if someone is reading this and wants to give it a try, I would recommend copying the [AutoScaling Group Data Source](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/data_source_aws_availability_zone.go) and test file, and modifying to do as I mentioned:

- grab name from config
- call `GetRestApis`
- loop until we find a match

The SDK includes `GetRestApisPages `:

- https://github.com/terraform-providers/terraform-provider-aws/blob/6097dd5c25599fe0ca9c2064be48dcdbce5f3256/vendor/github.com/aws/aws-sdk-go/service/apigateway/api.go#L6760-L6779

**Comment by bflad** [comment]
The `aws_api_gateway_rest_api` data source has been released in [version 1.15.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1150-april-18-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
